pandas/io/xml.py

```diff
@@ -0,0 +1,966 @@
+        validate xpath, names, parse and return specific nodes.
+        """
+
+        raise AbstractMethodError(self)
+
+    def _parse_nodes(self) -> List[Dict[str, Optional[str]]]:
+        """
+        Parse xml nodes.
+
+        This method will parse the children and attributes of elements
+        in xpath, conditionally for only elements, only attributes
+        or both while optionally renaming node names.
+
+        Raises
+        ------
+        ValueError
+            * If only elements and only attributes are specified.
+
+        Notes
+        -----
+        Namespace URIs will be removed from return node values.Also,
+        elements with missing children or attributes compared to siblings
+        will have optional keys filled withi None values.
+        """
+
+        raise AbstractMethodError(self)
+
+    def _validate_path(self) -> None:
+        """
+        Validate xpath.
+
+        This method checks for syntax, evaluation, or empty nodes return.
+
+        Raises
+        ------
+        SyntaxError
+            * If xpah is not supported or issues with namespaces.
+
+        ValueError
+            * If xpah does not return any nodes.
+        """
+
+        raise AbstractMethodError(self)
+
+    def _validate_names(self) -> None:
+        """
+        Validate names.
+
+        This method will check if names is a list-like and aligns
+        with length of parse nodes.
+
+        Raises
+        ------
+        ValueError
+            * If value is not a list and less then length of nodes.
+        """
+        raise AbstractMethodError(self)
+
+    def _preprocess_data(self, data):
+        """
+        Convert extracted raw data.
+
+        This method will return underlying data of extracted XML content.
+        The data either has a `read` attribute (e.g. a file object or a
+        StringIO/BytesIO) or is a string or bytes that is an XML document.
+        """
+        if hasattr(data, "read"):
+            data = data.read()
+
+        if not hasattr(data, "read") and isinstance(data, str):
+            data = io.StringIO(data)
+
+        if not hasattr(data, "read") and isinstance(data, bytes):
+            data = io.BytesIO(data)
+
+        return data
+
+    def _get_data_from_filepath(self, filepath_or_buffer):
+        """
+        Extract raw XML data.
+
+        The method accepts three input types:
+            1. filepath (string-like)
+            2. file-like object (e.g. open file object, StringIO)
+            3. XML bytes
+
+        This method turns (1) into (2) to simplify the rest of the processing.
+        It returns input types (2) and (3) unchanged.
+        """
+        filepath_or_buffer = stringify_path(filepath_or_buffer)
+        if (
+            not isinstance(filepath_or_buffer, str)
+            or is_url(filepath_or_buffer)
+            or is_fsspec_url(filepath_or_buffer)
+            or file_exists(filepath_or_buffer)
+        ):
+            self.handles = get_handle(
```

Comment:
Can you make the class a context manager (`__enter__` and `__exit__`) to call `self.handles.close()`. If the calling code can't use a context manager, then `try/finally` and implement `close()`.

Or better (if possible), let the function return the content, then you don't need to carry the file handles around.

Reply:
I implemented a `with` context for function to return value. This may even resolve lingering CI test failure of unclosed file for `io.TextIOWrapper` (interestingly only for `etree` and not `lxml` and only on the `Linux_py38_np_dev` run).

Reply:
running pytest with `-W default`  might help you find those `ResourceWarning`s